src/main.cpp

```diff
@@ -2363,11 +2363,128 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+        // so local code analysis tools don't think we have an unhandled case
+        // that could cause a crash:
+        if (block.vtx.empty()) {
+            return state.DoS(100, error("%s: first tx is not coinbase", __func__),
+                             REJECT_INVALID, "bad-cb-missing");
+        }
+        // Make sure there is at least *one* output in the coinbase which
+        // satisfies our spend criteria.
+        uint32_t idx;
+        const CTransaction& coinbaseTx = block.vtx.front();
+        for (idx = 0; idx < (uint32_t)coinbaseTx.vout.size(); ++idx) {
+            if (IsTriviallySpendable(coinbaseTx, idx, flags|SCRIPT_VERIFY_CLEANSTACK)) {
+                // Save the hash to the coin view.
+                view.SetFinalTx(coinbaseTx.GetHash());
+                // We have a winner, so stop looking.
+                break;
+            }
+        }
+        // Check if we found any suitable outputs.
+        if (idx == coinbaseTx.vout.size() || view.GetFinalTx().IsNull()) {
+            return state.DoS(100, error("%s: coinbase missing output for block-final tx", __func__),
+                             REJECT_INVALID, "bad-cb-missing-block-final-output");
+        }
+    }
+
+    else if (enforce_block_final) {
+        // The view contains the hash of the last block-final transaction, which
+        // necessarily *MUST* exist in the unspent txout database, since it
+        // occured in a prior block.
+        const CCoins* prev_final_tx = view.AccessCoins(view.GetFinalTx());
+        if (!prev_final_tx) {
+            // This state of affairs can potentially happen when performing an
+            // upgrade from a prior node version to one which supports
+            // validation of block-final transaction rules, but *after* rule
+            // activation. Normal tip validation in init.cpp will halt on this
+            // error and notify that the user that their block database is
+            // corrupted, which is fixed by starting with -reindex=1.
+            return state.DoS(100, error("%s: prior block-final tx hash not found; corruption likely!", __func__),
+                             REJECT_INVALID, "corrupt-db-no-blockfinal-hash", true);
+        }
+        // The output spent by the very first block-final transaction is
+        // generated in a coinbase at the point of activation, so it takes
+        // COINBASE_MATURITY blocks until that output matures and we're able to
+        // really start enforcing the block-final tx rules.
+        if (!prev_final_tx->IsCoinBase() || (pindex->nHeight - prev_final_tx->nHeight >= COINBASE_MATURITY)) {
+            // Make sure the block contains *at least* 2 transactions, the
+            // coinbase and the block-final transaction. This is not stricly
+            // necessary since coinbase can't have inputs, but for clarity we
+            // explicitly check and fail with the relevant error message.
+            if (block.vtx.size() < 2) {
```

Comment:
The transaction wouldn't be spendable in the next block due to maturity rules. It also couldn't include inputs from the previous block-final transaction. So the block-final transaction has to be NOT the coinbase (except for the initial output at activation which gets the chain going, which must be in the coinbase because it's the only thing we can assume a random miner has access to). So even if a block is empty, it needs to have two transactions, one the coinbase and one regular block-final transaction. Does that make sense, or did I misread what you meant?

Reply:
Yeah I think I forgot about the coinbase maturity rule, actually.